Dev lesson2

LiyaKarimova opened the pull request:
Домашнее задание №2

LiyaKarimova committed:
Исправление ошибок из первого дз

LiyaKarimova committed:
Перевела приложение на Annotation based

LiyaKarimova committed:
Подключила property file

LiyaKarimova committed:
Подключила property file

LiyaKarimova committed:
Дописала новый функционал

LiyaKarimova committed:
Юнит тест

LiyaKarimova committed:
Юнит тест

kataus commented on src/main/java/ru/otus/homework/Main.java:
```diff
@@ -1,12 +1,23 @@
+        AnnotationConfigApplicationContext context = new AnnotationConfigApplicationContext(Main.class);
+
+        TaskService taskService = context.getBean(TaskServiceImpl.class);
+        ResultInfoPrinter resultInfoPrinter = context.getBean(ResultInfoImpl.class);
```
Вся логика должны быть под контекстом. Снаружи только инициализация контекста и запуск 1 метода 1 бина

kataus commented on src/main/java/ru/otus/homework/dao/QuestionDaoImpl.java:
```diff
@@ -2,24 +2,27 @@
 public class QuestionDaoImpl implements QuestionDao {
 
-    private final CSVQuestionParser csvParser;
+    private final QuestionReaderImpl csvParser;
```
Вот именно такая реализация и никакая иная? По возможности не стоит привязываться к конкретной реализации, делегировать задачу выбора реализации Spring

kataus commented on src/main/java/ru/otus/homework/domain/ResultItem.java:
```diff
@@ -0,0 +1,15 @@
+@AllArgsConstructor
+public class ResultItem {
+
+    private final Question question;
```
А гетеры не нужны?

kataus commented on src/main/java/ru/otus/homework/exception/ParseException.java:
```diff
@@ -0,0 +1,12 @@
+package ru.otus.homework.exception;
+
+public class ParseException extends Exception {
```
Принято максимально быстро уходить от checkException к unchecked. Это и удобнее и упрощает разработку. Первоисточник мысли : https://docs.oracle.com/javase/tutorial/essential/exceptions/runtime.html

kataus commented on src/main/java/ru/otus/homework/exception/StringReaderException.java:
```diff
@@ -0,0 +1,8 @@
+package ru.otus.homework.exception;
+
+public class StringReaderException extends Exception {
```
Удобно использовать своё единое корневое исключение. Тогда при обработке можно будет легко отделить наши и не наши исключения

kataus commented on src/main/java/ru/otus/homework/service/QuestionIOService.java:
```diff
@@ -0,0 +1,29 @@
+@AllArgsConstructor
+@Slf4j
+@Service
+public class QuestionIOService {
```
Интерфейсы - очень сильный инструмент первичной абстракции кода. Их активное использование очень удобно и реально помогает строить гибкий код. Не стоит ими пренебрегать

kataus commented on src/main/resources/spring-context.xml:
```diff
@@ -3,18 +3,25 @@
-
-    <bean id="csvQuestionParser" class="ru.otus.homework.util.CSVQuestionParser">
-        <property name="fileName" value="questions.csv"></property>
+    <bean id = "stringReader" class="ru.otus.homework.util.CSVResourceStringReader">
```
Этот файл нам больше не нужен, далее используются аннотации

kataus commented on src/test/java/ru/otus/homework/QuestionReaderTest.java:
```diff
@@ -0,0 +1,38 @@
+
+    @BeforeEach
+    void prepare () {
+        stringReader = new CSVResourceStringReader("questions.csv");
```
Стоит замокать, по концепции unit- тестировая класс тестирует в изоляции, все внешние зависимости мокируются

LiyaKarimova committed:
Сделала запуск 1 методом 1 бина

LiyaKarimova committed:
1. Сделала всю логику под контекстом, сделала запуск приложения вызовом одного метода одного класса
2. Заменила инжекты классов инжектами интерфейсов
3. Добавила геттеры в классе ResultItem
4. Сделала свое корневое исключение, сделала extends собственных исключений от RuntimeException (вместо Exception)
5.Удалила файл spring-context.xml
6.Переделала unit test сервиса QuestionReader.

LiyaKarimova committed:
1. Сделала всю логику под контекстом, сделала запуск приложения вызовом одного метода одного класса
2. Заменила инжекты классов инжектами интерфейсов
3. Добавила геттеры в классе ResultItem
4. Сделала свое корневое исключение, сделала extends собственных исключений от RuntimeException (вместо Exception)
5.Удалила файл spring-context.xml
6.Переделала unit test сервиса QuestionReader.

kataus commented on src/main/java/ru/otus/homework/service/QuestionIOServiceImpl.java:
```diff
@@ -0,0 +1,28 @@
+    private final StringToAnswerParser stringToAnswerParser;
+
+
+    public void printQuestion (Question question) {
```
Не стоит пренебрегать @Override, очень удобная при чтении кода аннотация

kataus commented on src/main/java/ru/otus/homework/service/TaskServiceImpl.java:
```diff
@@ -0,0 +1,47 @@
+
+    }
+
+    public Answer doQuestion (Question question) {
```
А зачем public?

kataus commented on src/main/java/ru/otus/homework/util/Interfaces/ResultInfoPrinter.java:
```diff
@@ -0,0 +1,8 @@
+package ru.otus.homework.util.Interfaces;
```
1. Принято использовать только строчные наименования в пакетах
2. Интерфейсы первичны, потому их пакеты не ниже чем у реализаций

kataus commented on src/main/java/ru/otus/homework/util/CSVResourceStringReader.java:
```diff
@@ -0,0 +1,48 @@
+import java.util.List;
+
+@Component
+public class CSVResourceStringReader implements StringReader {
```
Это всё таки уже не утилитный класс, а вполне себе бин

LiyaKarimova committed:
Исправление ошибок после второй проверки дз
